Refactor the API of the matmul implementation

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76828
* #75197
* #75196
* #75195
* #75279
* __->__ #75194

Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

lezcano committed:
Refactor the API of the matmul implementation

Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

[ghstack-poisoned]

lezcano committed:
Update on "Refactor the API of the matmul implementation"


Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

[ghstack-poisoned]

lezcano committed:
Update on "Refactor the API of the matmul implementation"


Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

[ghstack-poisoned]

lezcano committed:
Update on "Refactor the API of the matmul implementation"


Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1759,25 +1758,26 @@ Tensor matmul(
   auto maybe_outnames = namedinference::compute_matmul_outnames(tensor1, tensor2);
-  auto result = at::native::matmul(c10::nullopt, tensor1, tensor2);
+  at::Tensor unused;
+  auto result = at::native::_matmul_impl(unused, tensor1, tensor2);
```
BTW empty initializer list should also work for triggering the undefined one

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1641,15 +1641,14 @@ The behavior depends on the dimensionality of the Tensors as follows:
-Tensor matmul(
-    c10::optional<Tensor> out_opt,
+Tensor _matmul_impl(
+    Tensor& out,
```
There actually isn't any reason this can't be a `const Tensor&`, which is better because you never directly overwrite the Tensor.

Lezcano replied:
I did this at some point, but it means that I'll need to use lots of const_casts in matmul. Is that preferrable?

ezyang replied:
sort of. The const casts are not good but eventually matmul will get structured-ified so that it doesn't take a mutable reference as argument and then you don't need to const cast. You can leave it like this.

ezyang reviewed (approved):
okey dokey

lezcano committed:
Update on "Refactor the API of the matmul implementation"


Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

[ghstack-poisoned]

lezcano committed:
Update on "Refactor the API of the matmul implementation"


Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

[ghstack-poisoned]

lezcano committed:
Update on "Refactor the API of the matmul implementation"


Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

[ghstack-poisoned]

lezcano committed:
Update on "Refactor the API of the matmul implementation"


Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1844,26 +1844,32 @@ void _fill_matrix_powers(Tensor& buffer, const Tensor& a, int num_matrices) {
-      buffer.select(0, 2), // out for a^2
+    // out for a^2
+    auto view_out = buffer.select(0, 2);
+    at::matmul_out(
```
why do we need to go through the dispatcher here instead of calling _matmul_impl?

Lezcano replied:
I thought that we should always go through the dispatcher? @albanD @zou3519

ngimel replied:
The "real" calls like `mm` have to go through dispatcher, but calling from one composite implicit function to another composite implicit function can be whatever (for all you know, `matmul` can be a helper function that you are calling from another one).

Lezcano replied:
Done!

lezcano committed:
Update on "Refactor the API of the matmul implementation"


Previously we used an odd overload using `c10::optional` to implement
the matmul logic of `matmul` and `matmul_out` simultaneously. This made
some functions (those in `linalg.matrix_exp`) call into this native::matmul
implementation, rather than going through the dispatcher.

In this PR we remove the use of `c10::optional` and rename the
implementation of matmul, to make sure that no one mistakenly calls it.

[ghstack-poisoned]

ngimel commented:
@pytorchbot merge this
